DeviceBridgeE2E/src/device-bridge.spec.ts

```diff
@@ -0,0 +1,205 @@
+import { setup, TestContext } from './utility/setup';
+import * as util from 'util';
+import * as fs from 'fs';
+import DeviceClient from './utility/device';
+import { assert } from 'console';
+import { sleep } from './utility/helpers'
+
+const test = anyTest as TestInterface<{
+    ctx: TestContext;
+    device: DeviceClient;
+    template: DeviceTemplate;
+}>;
+
+test.before(async t => {
+    t.context.ctx = await setup(t);
+
+    var deviceTemplate = JSON.parse(
+        await util.promisify(fs.readFile)(
+            './data/deviceTemplates/basic.json',
+            'utf-8'
+        )
+    );
+
+    t.context.template = deviceTemplate;
+
+    await t.context.ctx.publicAPI.createDeviceTemplate(t, deviceTemplate);
+
+    var device = await t.context.ctx.publicAPI.createDevice(t, {
+        id: v4(),
+        displayName: 'Basic Telemetry Device',
+        instanceOf: deviceTemplate.id,
+        simulated: false,
+    });
+
+    let creds: DeviceCredentials | undefined;
+    for (let i = 0; i < 5; i += 1) {
+        creds = await t.context.ctx.publicAPI.getDeviceCredentials(
+            t,
+            device.id
+        );
+        if (creds.symmetricKey) {
+            // Sometimes the credentials have no symmetric key. This is
+            // generally just a race condition.
+            t.log(`Received device credentials for device ${device.id}`);
+            break;
+        }
+
+        t.log(
+            `Received device credentials response with no symmetric key. Retrying...`
+        );
+        await new Promise(resolve => setTimeout(resolve, 5000));
+    }
+
+    if (!creds || !creds.symmetricKey) {
+        t.fail('invalid credentials');
+        return;
+    }
+
+    t.context.device = await DeviceClient.create(
+        device.id,
+        creds.idScope,
+        creds.symmetricKey.primaryKey
+    );
+});
+
+
+
+
+test.serial('Test device command callback', async t => {
+    // Create subscription to Azure Function
+    const callbackUrl = `${t.context.ctx.callbackUrl}&deviceId=${t.context.device.id}`;
+    await t.context.ctx.deviceBridgAPI.createCMDSubscription(t, t.context.device.id, callbackUrl)
+    
+    // Ensure get works
+    var response = await t.context.ctx.deviceBridgAPI.getCMDSubscription(t, t.context.device.id);
+    t.is(response.body.callbackUrl, callbackUrl)
+    t.is(response.body.status, "Running");
+    await t.context.ctx.publicAPI.executeCommand(t, t.context.device.id, "cmd");
+    var cmdInvocationValue = await t.context.ctx.deviceBridgAPI.getEcho(t, t.context.device.id);
+    var cmdInvocationValueBody = JSON.parse(cmdInvocationValue.body);
+    t.is("cmd", cmdInvocationValueBody.methodName);
+    t.is("DirectMethodInvocation", cmdInvocationValueBody.eventType);
+    t.is(t.context.device.id, cmdInvocationValueBody.deviceId);
+    await t.context.ctx.deviceBridgAPI.deleteCMDSubscription(t, t.context.device.id);
+});
+
+test.serial('Test device connection status callback', async t => {
```

Comment:
Can you move this one to be the first test? (seems a more intuitive order having the connection status test first)